crml/nft/src/lib.rs

```diff
@@ -974,6 +1001,130 @@ decl_module! {
+				}
+			}
+			// check user has the required funds to make this offer
+			let balance = T::MultiCurrency::free_balance(&origin, asset_id);
+			if let Some(balance_after_bid) = balance.checked_sub(amount) {
+				// TODO: review behaviour with 3.0 upgrade: https://github.com/cennznet/cennznet/issues/414
+				// - `amount` is unused
+				// - if there are multiple locks on user asset this could return true inaccurately
+				// - `T::MultiCurrency::reserve(origin, asset_id, amount)` should be checking this internally...
+				let _ = T::MultiCurrency::ensure_can_withdraw(&origin, asset_id, amount, WithdrawReasons::RESERVE, balance_after_bid)?;
+			}
+
+			// try lock funds
+			T::MultiCurrency::reserve(&origin, asset_id, amount)?;
+			TokenOffers::append(token_id, offer_id);
+			let new_offer = OfferType::<T::AccountId>::Simple(
+				SimpleOffer{
+					token_id,
+					asset_id,
+					amount,
+					buyer: origin.clone(),
+					marketplace_id,
+				}
+			);
+			Offers::<T>::insert(offer_id, new_offer);
+			NextOfferId::mutate(|i| *i += 1);
+
+			Self::deposit_event(RawEvent::OfferMade(offer_id, amount, asset_id, marketplace_id, origin));
+			Ok(())
+		}
+
+		/// Cancels an offer on a token
+		/// Caller must be the offer buyer
+		#[weight = 0]
+		fn cancel_offer (
+			origin,
+			offer_id: OfferId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			if let Some(offer_type) = Self::offers(offer_id) {
+				match offer_type {
+					OfferType::Simple(offer) => {
+						ensure!(offer.buyer == origin, Error::<T>::NotBuyer);
+						T::MultiCurrency::unreserve(&origin, offer.asset_id, offer.amount);
+						Offers::<T>::remove(offer_id);
+						TokenOffers::remove(offer.token_id);
+						Self::deposit_event(RawEvent::OfferCancelled(offer_id));
+						Ok(())
+					}
+					_ => Err(Error::<T>::OfferTypeNotImplemented.into())
+				}
+			} else {
+				Err(Error::<T>::InvalidOffer.into())
+			}
+		}
+
+		/// Accepts an offer on a token
+		/// Caller must be token owner
+		#[weight = 0]
+		fn accept_offer (
+			origin,
+			offer_id: OfferId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			if let Some(offer_type) = Self::offers(offer_id) {
+				match offer_type {
+					OfferType::Simple(offer) => {
+						let token_id = offer.token_id;
+						ensure!(Self::token_owner((token_id.0, token_id.1), token_id.2) == origin, Error::<T>::NoPermission);
+
+						let royalties_schedule = Self::check_bundle_royalties(&vec![token_id], offer.marketplace_id)?;
+
+						if let Err(err) = Self::process_token_payment(
+							&offer.buyer,
+							&origin,
+							offer.asset_id,
+							vec![offer.token_id],
+							offer.amount,
+							royalties_schedule,
+						) {
+							// offer settlement failed despite our prior validations.
+							// release winning bid funds
+							log!(error, "🃏 bid settlement failed: {:?}", err);
+							T::MultiCurrency::unreserve(&offer.buyer, offer.asset_id, offer.amount);
+
+						} else {
```

Comment:
simply take payment or fail should do
```suggestion
						Self::process_token_payment(
							&offer.buyer,
							&origin,
							offer.asset_id,
							vec![offer.token_id],
							offer.amount,
							royalties_schedule,
						)?;
```